### PR TITLE
[SPARK-7186] [SQL] Decouple internal Row from external Row

### PR DESCRIPTION
**Pull request by davies**
Currently, we use o.a.s.sql.Row both internally and externally. The external interface is wider than what the internal needs because it is designed to facilitate end-user programming. This design has proven to be very error prone and cumbersome for internal Row implementations.

As a first step, we create an InternalRow interface in the catalyst module, which is identical to the current Row interface. And we switch all internal operators/expressions to use this InternalRow instead. When we need to expose Row, we convert the InternalRow implementation into Row for users.

For all public API, we use Row (for example, data source APIs), which will be converted into/from InternalRow by CatalystTypeConverters.

For all internal data sources (Json, Parquet, JDBC, Hive), we use InternalRow for better performance, casted into Row in buildScan() (without change the public API). When create a PhysicalRDD, we cast them back to InternalRow.

cc @rxin @marmbrus @JoshRosen

### CODE REVIEW

**Comment by marmbrus** [comment]
Thanks for working on this huge change!  Should `InternalRow` live in catalyst with the other semi-private APIs?

**Comment by rxin** [comment]
How come Jenkins didn't print any meaning messages? cc @JoshRosen

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala [hunk 1]
```diff
@@ -17,6 +17,8 @@
 
 package org.apache.spark.sql.catalyst.expressions
 
+import org.apache.spark.sql.catalyst
```
Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: why not just import InternalRow?

**Comment by JoshRosen** [comment]
Jenkins, retest this please.

**Comment by JoshRosen** [comment]
I looked at the Jenkins configuration log and it looks like @andrewor14's credentials somehow auto-filled and overwrote the Jenkins GitHub token; Andrew and I were modifying the builder configurations this afternoon to attach `unit-tests.log` outputs to the builds as build artifacts (we'll email the dev list later with more details on this feature).

I've rolled back the configuration change so hopefully we'll see SparkQA posting soon.

For those with Jenkins admin access, see https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/jobConfigHistory/showDiffFiles?timestamp1=2015-06-12_09-30-32&timestamp2=2015-06-12_14-37-42

**Comment by rxin** [comment]
Going to merge this quickly since it conflicts with a lot of other patches.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala (hunk 1 above)
Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: This is done by Intellj.

Comment (marmbrus) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: Okay, but we aren't even consistent in our usage.  Can we remove this and just reference `InternalRow`?

Comment (rxin) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: +1 i think we should fix this

Comment (davies) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/Projection.scala]: Will do in a follow up PR. Please continue review this.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala [hunk 2]
```diff
@@ -68,26 +68,29 @@ case class HashOuterJoin(
...
+      joinedRow: JoinedRow,
+      rightIter: Iterable[InternalRow]): Iterator[InternalRow] = {
+    val ret: Iterable[InternalRow] = {
       if (!key.anyNull) {
         val temp = rightIter.collect {
           case r if boundCondition(joinedRow.withRight(r)) => joinedRow.copy()
         }
         if (temp.size == 0) {
-          joinedRow.withRight(rightNullRow).copy :: Nil
+          joinedRow.withRight(rightNullRow).copy.asInstanceOf[InternalRow] :: Nil
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: Is this cast needed?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/joins/HashOuterJoin.scala]: not needed, will remove it.

File: sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala [hunk 3]
```diff
@@ -195,6 +196,8 @@ abstract class BaseRelation {
    *  java.lang.String -> UTF8String
    *  java.lang.Decimal -> Decimal
    *
+   * If `needConversion` is `false`, buildScan() should return an [[RDD]] of [[InternalRow]]
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: So are we just relying on type erasure here to avoid class cast exceptions?

Comment (davies) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: If we move to let InternalRow no inherit from Row, maybe we could introduce another `internalBuildScan` for internal data sources, which will return `RDD[InternalRow]`, had not figured out all the details yet.

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/sources/interfaces.scala]: Yeah, I don't have a great suggestion either.  This just stuck out at the hackiest part to me.

**Comment by marmbrus** [comment]
Another nit:

```
[warn] /home/michael/spark/sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/ScalaReflectionSuite.scala:24: imported `InternalRow' is permanently hidden by definition of object InternalRow in package catalyst
```

**Pull request closed without merging**